net_if.c

```diff
@@ -0,0 +1,304 @@
+		close(net_sock);
+		errno = err;
+		return -1;
+	}
+	return net_sock;
+}
+
+static int send_stats_req(int net_sock)
+{
+	int err;
+	struct stats {
+		struct nlmsghdr nl_hdr;
+		struct if_stats_msg if_stm;
+	} req;
+
+	memset(&req, 0, sizeof(req));
+	req.nl_hdr.nlmsg_type = RTM_GETSTATS;
+	req.nl_hdr.nlmsg_len = sizeof(req);
+	req.nl_hdr.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
+	req.nl_hdr.nlmsg_seq = 0;
+	req.nl_hdr.nlmsg_pid = getpid();
+	
+	req.if_stm.filter_mask = IFLA_STATS_LINK_64;
+	req.if_stm.family = AF_UNSPEC;
+	req.if_stm.ifindex = 0;
+
+	err = send(net_sock, &req, req.nl_hdr.nlmsg_len, 0);
+	if (err == -1) {
+		err = errno;
+		eprintf("Error in request send(), errno %d\n", err);
+		errno = err;
+		return -1;
+	}
+	return 0;
+}
+
+static int parse_attr(struct rtattr *at[], struct rtattr *rta, int len)
+{
+	memset(at, 0, sizeof(struct rtattr *) * (IFLA_STATS_MAX + 1));
+	while (RTA_OK(rta, len)) {
+		if (rta->rta_type <= IFLA_STATS_MAX)
+			at[rta->rta_type] = rta;
+		rta = RTA_NEXT(rta, len);
+	}
+	return 0;
+}
+
+static int get_if_len(int net_sock)
+{
+	int recvl, err;
+	struct msghdr msg;
+	struct iovec iov;
+	struct sockaddr_nl nladdr;
+
+	msg.msg_name = &nladdr;
+	msg.msg_namelen = sizeof(nladdr);
+	iov.iov_base = NULL;
+	iov.iov_len = 0;
+
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	recvl = recvmsg(net_sock, &msg, MSG_PEEK | MSG_TRUNC);
+	if (recvl < 0) {
+		err = errno;
+		eprintf("Error in get length recvmsg(), errno %d\n", err);
+		errno = err;
+		return -1;
+	}
+#ifdef PRINTLOGS
+	printf("received len %d\n",recvl);
+#endif
+	return recvl;
+}
+
+static int get_max_ifaces(int len)
+{
+	int header_size, max_if;
+
+	header_size = (sizeof(struct nlmsghdr)) +
+			(sizeof(struct if_stats_msg)) +
+			(sizeof(struct rtnl_link_stats64));
+#ifdef PRINTLOGS
+	printf("header_size %d len %d\n",header_size, len);
+#endif
+	max_if = len/header_size;
+	return max_if;
+}
+
+static int handle_stats_resp(int net_sock, void **out)
+{
+	int err, recvl, len, max_ifs, if_size;
+	int ifind = 0, total_if_size = 0;
+	struct iovec iov;
+	struct msghdr msg;
+	struct nlmsghdr *r;
+	struct nlmsgerr *nlmsg_err;
+	struct if_stats_msg *ifsm;
+	struct rtattr *at[IFLA_STATS_MAX+1];
+	struct rtattr *rta;
+	struct ifstats *ifaces = NULL, *ifs = NULL;
+	struct sockaddr_nl nladdr;
+
+	msg.msg_name = &nladdr;
+	msg.msg_namelen = sizeof(nladdr);
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	while(1) {
+		len = get_if_len(net_sock);
+		if (len == -1) {
+			if (ifaces)
+				free(ifaces);
+			return -1;
+		}
+		char *buf = (char *)malloc(len);
+		if (!buf) {
+			if (ifaces)
+				free(ifaces);
+			return -1;
+		}
+		iov.iov_base = buf;
+		iov.iov_len = len;
+		max_ifs = get_max_ifaces(len);
+		if_size = max_ifs * sizeof(struct ifstats);
+		total_if_size += if_size;
+#ifdef PRINTLOGS
+		printf("max_ifs %d, if_size %d total_if_size %d\n",max_ifs, if_size, total_if_size);
+#endif
+		if (ifaces == NULL) {
+			ifaces = (struct ifstats *) malloc(total_if_size);
```

Comment:
Thanks, this looks good, one last comment though.  Alex would yell at me if I did not point out that realloc(NULL, size) works as malloc.  So you don't need this whole if (ifaces == NULL) case is not needed, you can just always do tmp = realloc(ifaces, total_if_size); and on the first run, ifaces will be NULL, which is fine.